pkg/controller/pinned-image-set/pinned_image_set_controller.go

```diff
@@ -0,0 +1,702 @@
+		if err != nil {
+			return err
+		}
+
+		imageSetTmp := newImageSet.DeepCopy()
+		// Only append the mc name if it is already not in the list of finalizers.
+		// When we update an existing pinned image set, the generation number increases causing
+		// a resync to happen. When this happens, the mc name is the same, so we don't
+		// want to add duplicate entries to the list of finalizers.
+		if !ctrlcommon.InSlice(mc.Name, imageSetTmp.Finalizers) {
+			imageSetTmp.Finalizers = append(imageSetTmp.Finalizers, mc.Name)
+		}
+
+		modJSON, err := json.Marshal(imageSetTmp)
+		if err != nil {
+			return err
+		}
+
+		patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(curJSON, modJSON, curJSON)
+		if err != nil {
+			return err
+		}
+		return ctrl.patch(imageSet.Name, patch)
+	})
+}
+
+func (ctrl *Controller) popFinalizer(imageSet *mcfgv1.PinnedImageSet) error {
+	return retry.RetryOnConflict(ctrl.waitBackOff, func() error {
+		newimageSet, err := ctrl.imageSetLister.Get(imageSet.Name)
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		curJSON, err := json.Marshal(newimageSet)
+		if err != nil {
+			return err
+		}
+
+		imageSetTmp := newimageSet.DeepCopy()
+		imageSetTmp.Finalizers = append(imageSet.Finalizers[:0], imageSet.Finalizers[1:]...)
+
+		modJSON, err := json.Marshal(imageSetTmp)
+		if err != nil {
+			return err
+		}
+
+		patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(curJSON, modJSON, curJSON)
+		if err != nil {
+			return err
+		}
+		return ctrl.patch(imageSet.Name, patch)
+	})
+}
+
+func (ctrl *Controller) patch(name string, patch []byte) error {
+	_, err := ctrl.mcfgClient.MachineconfigurationV1().PinnedImageSets().Patch(context.TODO(), name, types.MergePatchType, patch, metav1.PatchOptions{})
+	return err
+}
+
+func createTOMLgeneratedConfigFile(path string, tomlConf interface{}) (generatedConfigFile, error) {
+	var newData bytes.Buffer
+	encoder := toml.NewEncoder(&newData)
+	if err := encoder.Encode(tomlConf); err != nil {
+		return generatedConfigFile{}, fmt.Errorf("error encoding toml for CRIO drop-in files: %w", err)
+	}
+	return generatedConfigFile{filePath: path, data: newData.Bytes()}, nil
+}
+
+type tomlConfigCRIOPinnedImages struct {
+	Crio struct {
+		Image struct {
+			PinnedImages []mcfgv1.PinnedImageRef `toml:"pinned_images,omitempty"`
+		} `toml:"image"`
+	} `toml:"crio"`
+}
+
+// generatedConfigFile is a struct that holds the filepath and data of the various configs
+// Using a struct array ensures that the order of the ignition files always stay the same
+// ensuring that double MCs are not created due to a change in the order
+type generatedConfigFile struct {
+	filePath string
+	data     []byte
+}
+
+// getMachineConfigKey returns the managed key for the machine config
+func getMachineConfigKey(pool *mcfgv1.MachineConfigPool, client mcfgclientset.Interface, imageSetOrig *mcfgv1.PinnedImageSet) (string, error) {
```

Comment:
I think this will work like the crio/kubelet configuration rendering, meaning that custom pool config > worker pool config now (but if you don't define a pinned image set for, say, your infra node, it will inherit worker configs and still try to pull as if it was a worker).

That's probably the expected behaviour but wanted to check explicitly

Reply:
Right since today. configs are deployed on a pool level I don't feel it makes sense for this controller to act in a different way. My understanding is that you can create a custom pool for dedicated to a certain purpose "infra"? In that case the config could be deployed to only those nodes which are pool members.